http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/Boot.scala

```diff
@@ -332,34 +333,52 @@ object Boot extends IOApp {
-      retryPolicy = RetryPolicy[F](RetryPolicy.exponentialBackoff(30 seconds, 5))
-
-      sslContext <- Resource.eval(SslContextReader.getSSLContext())
-      httpClientWithCustomSSL <- blaze.BlazeClientBuilder[F](blockingEc, Some(sslContext)).resource
-      clientWithRetryWithCustomSSL = Retry(retryPolicy)(httpClientWithCustomSSL)
-      clientWithRetryAndLogging = Http4sLogger[F](logHeaders = true, logBody = false)(clientWithRetryWithCustomSSL)
 
       // This is for sending custom metrics to stackdriver. all custom metrics starts with `OpenCensus/leonardo/`.
       // Typing in `leonardo` in metrics explorer will show all leonardo custom metrics.
       // As best practice, we should have all related metrics under same prefix separated by `/`
       implicit0(openTelemetry: OpenTelemetryMetrics[F]) <- OpenTelemetryMetrics
         .resource[F](applicationConfig.leoServiceAccountJsonFile, applicationConfig.applicationName, blocker)
 
-      // Note the Sam client intentionally doesn't use clientWithRetryAndLogging because the logs are
-      // too verbose. We send OpenTelemetry metrics instead for instrumenting Sam calls.
-      samDao = HttpSamDAO[F](clientWithRetryWithCustomSSL, httpSamDaoConfig, blocker)
+      // Set up database reference
       concurrentDbAccessPermits <- Resource.eval(Semaphore[F](dbConcurrency))
       implicit0(dbRef: DbReference[F]) <- DbReference.init(liquibaseConfig, concurrentDbAccessPermits, blocker)
-      runtimeDnsCache = new RuntimeDnsCache(proxyConfig, dbRef, runtimeDnsCacheConfig, blocker)
-      kubernetesDnsCache = new KubernetesDnsCache(proxyConfig, dbRef, kubernetesDnsCacheConfig, blocker)
-      welderDao = new HttpWelderDAO[F](runtimeDnsCache, clientWithRetryAndLogging)
-      dockerDao = HttpDockerDAO[F](clientWithRetryAndLogging)
-      jupyterDao = new HttpJupyterDAO[F](runtimeDnsCache, clientWithRetryAndLogging)
-      rstudioDAO = new HttpRStudioDAO(runtimeDnsCache, clientWithRetryAndLogging)
+
+      // Set up DNS caches
+      hostToIpMapping <- Resource.eval(Ref.of(Map.empty[Host, IP]))
+      proxyResolver = new ProxyResolverInterp(hostToIpMapping)
+      runtimeDnsCache = new RuntimeDnsCache(proxyConfig, dbRef, runtimeDnsCacheConfig, hostToIpMapping, blocker)
+      kubernetesDnsCache = new KubernetesDnsCache(proxyConfig,
+                                                  dbRef,
+                                                  kubernetesDnsCacheConfig,
+                                                  hostToIpMapping,
+                                                  blocker)
+
+      // Set up SSL context and http clients
+      retryPolicy = RetryPolicy[F](RetryPolicy.exponentialBackoff(30 seconds, 5))
+      sslContext <- Resource.eval(SslContextReader.getSSLContext())
+      httpClient <- blaze
+        .BlazeClientBuilder[F](blockingEc, Some(sslContext))
+        .withCustomDnsResolver(proxyResolver.resolveHttp4s)
```

Comment:
we only need to use custom dns resolver for proxy requests right?

Reply:
Hm that's true I guess, though I made the resolver fall back to default resolution if the host is not in the dns cache. But maybe cleaner would be a separate client for proxy requests.

Reply:
ah I see...that makes sense...maybe just add a comment about this?

Reply:
Sure sounds good